### ISSUE TITLE
Add model variations for Megalodon innovations

### ISSUE THREAD

**Issue by gkielian** [ReaLLMASIC/nanoGPT#354]
From the paper:

- Complex exponential moving average
- timestep norm
- normalized attention
- two-hop residual

Reference:
https://arxiv.org/html/2404.08801v2

For consideration:

```
--- a/model.py
+++ b/model.py
@@ -1,5 +1,7 @@
 import math
+import numpy as np
 import torch
+import torch.nn.functional as F
 import torch.nn as nn
 from torch.nn import functional as F
 
@@ -19,6 +21,69 @@
         self.dropout = nn.Dropout(config.dropout)
 
     def forward(self, x):
+        return self.dropout(self.ln(self.c_fc(x)))
+
+class ComplexLinear(nn.Module):
+    def __init__(self, in_features, out_features):
+        super().__init__()
+        self.real = nn.Linear(in_features, out_features)
+        self.imag = nn.Linear(in_features, out_features)
+
+    def forward(self, x):
+        return torch.complex(self.real(x), self.imag(x))
+
+class TimestepNormalization(nn.Module):
+    def __init__(self, num_features, num_groups=32, eps=1e-5):
+        super().__init__()
+        self.num_groups = num_groups
+        self.eps = eps
+        self.weight = nn.Parameter(torch.zeros(num_features))
+        self.bias = nn.Parameter(torch.zeros(num_features))
+
+    def forward(self, x):
+        # x: (B, T, C)
+        B, T, C = x.size()
+        G = self.num_groups
+        
+        # Reshape to (B, G, T, C//G)
+        x = x.view(B, G, T, C // G)
+
+        # Cumulative sum for mean and variance calculation
+        cum_sum = torch.cumsum(x, dim=2)
+        cum_squared_sum = torch.cumsum(x * x, dim=2)
+
+        # Calculate mean and variance
+        count = torch.arange(1, T + 1).view(1, 1, T, 1).to(x.device)
+        mean = cum_sum / count
+        variance = (cum_squared_sum - 2 * mean * cum_sum + count * mean * mean) / count
+        
+        # Normalize
+        x = (x - mean) / torch.sqrt(variance + self.eps)
+
+        # Reshape back to (B, T, C)
+        x = x.view(B, T, C)
+        
+        # Apply learnable affine parameters with +1 reparameterization
+        x = (self.weight + 1) * x + self.bias
+
+        return x
+
+class CEMA(nn.Module):
+    def __init__(self, dim, hidden_dim, chunk_size):
+        super().__init__()
+        self.alpha = nn.Parameter(torch.rand(dim, hidden_dim))
+        self.delta = nn.Parameter(torch.rand(dim, hidden_dim))
+        self.omega = nn.Parameter(torch.rand(dim))
+        self.proj = ComplexLinear(dim, hidden_dim)
+        self.proj_out = ComplexLinear(hidden_dim, dim)
+        self.chunk_size = chunk_size
+
+    def forward(self, x):
+        # x: (B, T, C)
+        B, T, C = x.size()
+
+        # Pad sequence to be multiple of chunk size
+        if T % self.chunk_size != 0:
+            pad_len = self.chunk_size - (T % self.chunk_size)
+            x = F.pad(x, (0, 0, 0, pad_len))
+            _, T, _ = x.size()
+
+        # Reshape to (B, num_chunks, chunk_size, C)
+        x = x.view(B, T // self.chunk_size, self.chunk_size, C)
+
+        # Apply complex linear projection
+        u = self.proj(x)  # (B, num_chunks, chunk_size, hidden_dim)
+
+        # Compute theta
+        h = u.size(-1)
+        theta = torch.einsum("bth,d->bdth", torch.arange(1, self.chunk_size + 1).float().to(x.device), self.omega)
+        theta = theta.unsqueeze(-1).repeat(1, 1, 1, h)
+        k = torch.arange(1, h + 1).view(1, 1, 1, h).to(x.device)
+        theta = 2 * np.pi * k * theta / h
+
+        # Compute exponential terms
+        exp_alpha_theta = torch.exp(torch.complex(self.alpha, torch.zeros_like(self.alpha))) * torch.exp(torch.complex(torch.zeros_like(theta), theta))
+        exp_delta_theta = torch.exp(torch.complex(self.delta, torch.zeros_like(self.delta))) * torch.exp(torch.complex(torch.zeros_like(theta), theta))
+
+        # Initialize hidden state
+        h_prev = torch.zeros_like(u[:, :, 0, :])
+
+        h_list = []
+        for t in range(self.chunk_size):
+            h_t = exp_alpha_theta[:, None, :, :] * u[:, :, t, :] + (1 - exp_alpha_theta[:, None, :, :] * exp_delta_theta[:, None, :, :]) * h_prev
+            h_list.append(h_t)
+            h_prev = h_t
+
+        h = torch.stack(h_list, dim=2)
+        y = self.proj_out(h).real  # (B, num_chunks, chunk_size, C)
+
+        # Reshape back to (B, T, C)
+        y = y.view(B, T, C)
+
+        # Remove padding
+        y = y[:, :T - pad_len, :]
+
         return y
 
 class CausalSelfAttention(nn.Module):
@@ -26,28 +91,36 @@
     def __init__(self, config):
         super().__init__()
         assert config.n_embd % config.n_head == 0
-        # key, query, value projections for all heads, but in a batch
-        self.c_attn = nn.Linear(config.n_embd, 3 * config.n_embd, bias=config.bias)
-        # output projection
-        self.c_proj = nn.Linear(config.n_embd, config.n_embd, bias=config.bias)
+        # CEMA projections
+        self.cema = CEMA(config.n_embd, config.n_embd, config.chunk_size)
+
+        # Shared representation projection
+        self.c_z = nn.Linear(config.n_embd, config.z_dim, bias=config.bias)
+        self.norm_z = nn.LayerNorm(config.z_dim)  # Normalize Z
+
+        # key, query, value projections
+        self.q_scalar = nn.Parameter(torch.zeros(config.z_dim))
+        self.q_offset = nn.Parameter(torch.zeros(config.z_dim))
+        self.k_scalar = nn.Parameter(torch.zeros(config.z_dim))
+        self.k_offset = nn.Parameter(torch.zeros(config.z_dim))
+        self.v_proj = nn.Linear(config.n_embd, config.v_dim, bias=config.bias)
+
+        # Output projection
+        self.c_proj = nn.Linear(config.v_dim, config.n_embd, bias=config.bias)
+
         # regularization
         self.attn_dropout = nn.Dropout(config.attn_dropout)
         self.resid_dropout = nn.Dropout(config.resid_dropout)
         self.n_head = config.n_head
         self.n_embd = config.n_embd
+        self.z_dim = config.z_dim
+        self.v_dim = config.v_dim
         self.dropout = config.dropout
-        # flash attention make GPU go brrrrr but support is only in PyTorch >= 2.0
-        self.flash = hasattr(torch.nn.functional, 'scaled_dot_product_attention')
-        if not self.flash:
-            print("WARNING: using slow attention. Flash Attention requires PyTorch >= 2.0")
-            # causal mask to ensure that attention is only applied to the left in the input sequence
-            self.register_buffer("bias", torch.tril(torch.ones(config.block_size, config.block_size))
-                                        .view(1, 1, config.block_size, config.block_size))
+        self.chunk_size = config.chunk_size
 
     def forward(self, x):
         B, T, C = x.size() # batch size, sequence length, embedding dimensionality (n_embd)
 
-        # calculate query, key, values for all heads in batch and move head forward to be the batch dim
+        # CEMA
+        x_cema = self.cema(x)
+
+        # Shared representation
+        z = self.c_z(x_cema)
+        z_norm = self.norm_z(z)  # Apply normalization to Z
+
+        # Compute query, key, value with normalization
+        q = (self.q_scalar + 1) * z_norm + self.q_offset  # (B, T, z_dim)
+        k = (self.k_scalar + 1) * z_norm + self.k_offset  # (B, T, z_dim)
+        v = self.v_proj(x)  # (B, T, v_dim)
+
+        # Split heads
+        q = q.view(B, T, self.n_head, self.z_dim // self.n_head).transpose(1, 2)  # (B, nh, T, z_dim//nh)
+        k = k.view(B, T, self.n_head, self.z_dim // self.n_head).transpose(1, 2)  # (B, nh, T, z_dim//nh)
+        v = v.view(B, T, self.n_head, self.v_dim // self.n_head).transpose(1, 2)  # (B, nh, T, v_dim//nh)
+
+        # Split into chunks
+        if T % self.chunk_size != 0:
+            pad_len = self.chunk_size - (T % self.chunk_size)
+            q = F.pad(q, (0, 0, 0, pad_len))
+            k = F.pad(k, (0, 0, 0, pad_len))
+            v = F.pad(v, (0, 0, 0, pad_len))
+            _, _, T, _ = q.size()
+        q = q.view(B, self.n_head, T // self.chunk_size, self.chunk_size, self.z_dim // self.n_head)
+        k = k.view(B, self.n_head, T // self.chunk_size, self.chunk_size, self.z_dim // self.n_head)
+        v = v.view(B, self.n_head, T // self.chunk_size, self.chunk_size, self.v_dim // self.n_head)
+
+        # Compute attention scores for each chunk
+        att_list = []
+        for i in range(T // self.chunk_size):
+            qi = q[:, :, i, :, :]
+            ki = k[:, :, i, :, :]
+            vi = v[:, :, i, :, :]
+            att = (qi @ ki.transpose(-2, -1)) * (1.0 / math.sqrt(ki.size(-1)))  # (B, nh, T//cs, cs) @ (B, nh, cs, T//cs) -> (B, nh, T//cs, T//cs)
+            att = F.softmax(att, dim=-1)
+            att = self.attn_dropout(att)
+            out_i = att @ vi  # (B, nh, T//cs, T//cs) @ (B, nh, T//cs, v_dim//nh) -> (B, nh, T//cs, v_dim//nh)
+            att_list.append(out_i)
+        out = torch.cat(att_list, dim=2)  # (B, nh, T, v_dim//nh)
+
+        # Remove padding
+        out = out[:, :, :T - pad_len, :]
+
         # causal self-attention; Self-attend: (B, nh, T, hs) x (B, nh, hs, T) -> (B, nh, T, T)
-        if self.flash:
-            # efficient attention using Flash Attention CUDA kernels
```
-            y = torch.nn.functional.scaled_dot_product_attention(q, k, v, attn_mask=None, dropout_p=self.dropout if self.training else 0, is_causal=True)
-        else:
-            # manual implementation of attention
-            att = (q @ k.transpose(-2, -1)) * (1.0 / math.sqrt(k.size(-1)))
-            att = att.masked_fill(self.bias[:,:,:T,:T] == 0, float('-inf'))
-            att = F.softmax(att, dim=-1)
-            att = self.attn_dropout(att)
-            y = att @ v # (B, nh, T, T) x (B, nh, T, hs) -> (B, nh, T, hs)
-        y = y.transpose(1, 2).contiguous().view(B, T, C) # re-assemble all head outputs side by side
+        out = out.transpose(1, 2).contiguous().view(B, T, self.v_dim)  # Re-assemble all head outputs side by side
